### PR TITLE
Create coverage report for GNU tests

### LINKED ISSUE #2946: Add a new job in the CI to compute the code coverage of this project using the GNU testsuite
Currently, the code coverage is computed based on the unit & integration tests of this project, it isn't using the GNU testsuite.

It would be interesting to build rust/coreutils with coverage support and then run the GNU testsuite instead of ours.

Now, results should be probably pushed as a separate project (so, not https://app.codecov.io/gh/uutils/coreutils)

The gnu testsuite execution code is in:
https://github.com/uutils/coreutils/blob/main/.github/workflows/GnuTests.yml
and the coverage:
https://github.com/uutils/coreutils/blob/main/.github/workflows/CICD.yml#L788

### PR DESCRIPTION
**Pull request by xxyzz**
- It seems codecov-action can only upload reports to the same project.
- The code runs slower with coverage report enabled, so I extend the timeout limit to four hours. 
- codecov-action@v1 is deprecated, I didn't update the action in CICD.yml in case it breaks something.
- The origin windows coverage report action failed on my forked repo sometimes, but I don't think it's caused by my changes.
- Install valgrind  and libexpect-perl for tests that require it.

Closes #2946.

### CODE REVIEW

**Comment by sylvestre** [comment]
excellent stuff, thanks :)
I was more thinking at a new CI job then extend "Run GNU tests" 

What about "Run GNU tests" and "Run GNU tests with coverage" ?

**Comment by rivy** [comment]
I very much appreciate this effort, but I'm concerned about adding this workflow to the usual CI runs.
The workflow triples the CI completion time from (an already too long) 1 hour to 3 hours.

And, in looking at the [coverage diff](https://codecov.io/gh/uutils/coreutils/commit/a29e379dd748f5d0579e57ebd9e6a8984e059cf7), I don't see much of a coverage change either... only a few lines here with lots of them being changes to coverage of help message lines.

Maybe only enable it for merges to 'main'?

**Comment by xxyzz** [comment]
I think [this link](https://app.codecov.io/gh/uutils/coreutils/compare/3045/changes) shows the correct coverage change of this pull request.

**Comment by sylvestre** [comment]
Hmm, if I understand it correctly, it is adding on top of the current one ?
it isn't a separate result?

**Comment by rivy** [comment]
> I think [this link](https://app.codecov.io/gh/uutils/coreutils/compare/3045/changes) shows the correct coverage change of this pull request.

@xxyzz , That's definitely a much better gain (on the order of ~500 net line coverage gain)!

Thanks for showing me that comparison mode.

So, any ideas about how can we add this in a way that doesn't severely slow the CI for regular PRs?
An hour seems manageable, if only barely, for the current commit cadence, but three seems too far out.

**Comment by xxyzz** [comment]
I suspect the test runs slower because optimizations are disabled. I already let the new job only runs on the main branch in the last commit as you suggested.

**Comment by sylvestre** [comment]
I approved earlier a patch from @rivy before
it caused conflicts here :( sorry
could you please fix them?
thanks

**Comment by rivy** [comment]
@xxyzz , I'm happy to fix the merge conflicts (especially since I caused them ☺️).

I want to merge some maintenance work on the configuration and 'util' scripts first, then I'll rebase your additions on top and push them back here. Ok?

**Comment by xxyzz** [comment]
Alright, thanks for solving the conflicts.

**Comment by rivy** [comment]
@xxyzz , I've rebased this on top of #3090, preserving your intent as I could see it.
When #3090 is merged, I'll update this commit set for you to review and revise.

File: .github/workflows/GnuTests.yml [hunk 1]
```diff
@@ -137,11 +137,13 @@ jobs:
         name: "${{ steps.summary.outputs.HASH }}"
         path: "${{ steps.vars.outputs.TEST_SUMMARY_FILE }}"
     - name: Reserve test results summary
+      if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
       uses: actions/upload-artifact@v2
       with:
         name: test-summary
         path: "${{ steps.vars.outputs.TEST_SUMMARY_FILE }}"
     - name: Reserve test logs
+      if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
```
Comment (rivy) [inline .github/workflows/GnuTests.yml:146]: Same as above... please remove.

File: .github/workflows/GnuTests.yml [hunk 2]
```diff
@@ -137,11 +137,13 @@ jobs:
         name: "${{ steps.summary.outputs.HASH }}"
         path: "${{ steps.vars.outputs.TEST_SUMMARY_FILE }}"
     - name: Reserve test results summary
+      if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
```
[the remark this answers is not in the document]
Comment (xxyzz) [reply .github/workflows/GnuTests.yml]: I though if the job runs the second time in a pull request and upload it's result, it'll compare with the previous result from the same pull request but not the main branch result.

[the remark this answers is not in the document]
Comment (rivy) [reply .github/workflows/GnuTests.yml]: No, the reference test results are downloaded from last completed GnuTests workflow on the default branch.
A testing or merging PR shouldn't change that download.
And, there shouldn't be a need to re-upload another result set.

**Comment by sylvestre** [comment]
let's see :)
many thanks

**Comment by xxyzz** [comment]
I should run `sudo` after the end of the shell file... There are many permission errors in the log: https://github.com/xxyzz/coreutils/runs/5180735222?check_suite_focus=true#step:8:1688

**Comment by sylvestre** [comment]
@xxyzz quick question, how can I use codecov to compare the gnu vs rust testsuite ? :)

**Comment by sylvestre** [comment]
I noticed that you have
```
        flags: gnutests
        name: gnutests
```
but I don't see them on https://app.codecov.io/gh/uutils/coreutils

**Comment by xxyzz** [comment]
The GNU tests report only created when commits pushed to the main branch, so only the merge commits have it. Codecov says "Flags have been temporarily removed" in the [compare page overview tab](https://codecov.io/gh/uutils/coreutils/compare/775588fb18be0989cb87f047544dde797aa2253a...38bcafb7c298cf2ac16ebf2fcdb514def4f266c2), but the [app subdomain](https://app.codecov.io/gh/uutils/coreutils/compare/775588fb18be0989cb87f047544dde797aa2253a...38bcafb7c298cf2ac16ebf2fcdb514def4f266c2/overview) still has it. And as the notice said, [flags can be used on files](https://codecov.io/gh/uutils/coreutils/compare/775588fb18be0989cb87f047544dde797aa2253a...38bcafb7c298cf2ac16ebf2fcdb514def4f266c2/changes).

**Comment by sylvestre** [comment]
thanks again :)

One more question @xxyzz 
it seems that the coverage diff analysis in PR is using GNU's as comparison
does it ring a bell ?
```
 codecov/project Successful in 1s — 69.31% (-4.34%) compared to f11ac4b 
```

**Comment by xxyzz** [comment]
f11ac4b0763c48ad1b3c8192d587eaa8cb218a53 is a merge commit, so it has one more GNU coverage report then a PR. Now the GNU coverage job takes around one hour to finish, maybe it's time to enable it in pull requests so it can be compared?

**Comment by sylvestre** [comment]
Sounds good :)

**Pull request merged**